enhancements/insights/insights_gateway.md

```diff
@@ -0,0 +1,199 @@
+---
+title: insights-gateway
+authors:
+  - "@iNecas"
+reviewers:
+  - "@jhjaggars"
+  - "@chambridge"
+  - "@martinkunc"
+  - "@mfojtik"
+approvers:
+  - "@mfojtik"
+  - "@smarterclayton"
+creation-date: 2020-08-20
+last-updated: 2020-08-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Insights Gateway
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Insights-operator provides a proxy/gateway to provide a secure yet simple
+way for additional components to send payloads from the OCP cluster to
+cloud.redhat.com. This reduces complexity on user as well as developer
+side and adds additional control over the data sent to cloud.redhat.com.
+
+## Motivation
+
+In the last couple of releases, we've seen multiple new services build in 
+cloud.redhat.com (c.r.c.) providing additional insights from various areas,
+each needing additional data to be sent from the OCP clusters, such as:
+
+- [cost management](https://github.com/project-koku/korekuta-operator)
+- [subscription watch](https://github.com/chambridge/subscription-watch-operator)
+- [marketplace](https://github.com/redhat-marketplace/redhat-marketplace-operator)
+
+With more managed services, this list is expected to grow. Currently, every service
+provides an OLM-managed operator that collects and sends the required data.
+The data are eventually sent to `https://cloud.redhat.com/api/ingress/v1/upload`
+
+One of the problems is that for each of the operators, one needs to configure
+correct credentials to authenticate the cluster against the remote resource
+(example in [cost management documentation](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.5/html/getting_started_with_cost_management/assembly_adding_sources_cost#configuring_cost_mgmt-operator)). This leads to:
+
+- increased complexity for the user to set up this components
+- the need for each component to implement their own way to send the data 
+- number of places the credentials to cloud.redhat.com are stored
+- no single place to audit and opt-out form the payloads being sent to c.r.c.
+
+Given insights-operator is already part of the OCP cluster and has proper credentials
+configured already as part of the installation, we're proposing for it to expose
+a proxy for other components to be able to send the payloads on their behalf.
```

Comment:
Hrm - I would probably prefer these components remain decoupled architecturally.  Insights is providing support data about the cluster - it really is not a generic operator (generic operators have a high bar).

There must be an extremely strong reason to expose a new generic api - it would probably be a better start to expose a library those operators can use rather than trying to provide a service.  The justifications over providing a library would have to be very compelling.

Reply:
Looping @jhjaggars and especially @chambridge to perhaps fill in some gaps around requirements/justification and more thoughts on the library alternative.

Few things regarding the library approach:
1. some targeted operators are implemented with ansible (https://github.com/project-koku/korekuta-operator/blob/5fe7938ee87ca08dec6181cc0e37083979def62c/roles/collect/tasks/main.yml#L339): the library doesn't seem to help much with that
2. with the library approach, the operator would need to have access to secrets in `openshift-config` namespace: would we want to allow this or minimize the access to that resource? Or would it mean the credentials need to be handled differently. The gateway approach limits the access to the upload only: nothing else exposed.
3. the library still doesn't allow the single place for overview on what data are flowing to c.r.c.
4. additional requirements around networking we've already seen and solved in insights-operator (such as service-specific proxy as implemented in https://github.com/openshift/insights-operator/commit/65e218343c05ed4ba98664232f893469f7583757) would need to be implemented again or left as gaps.

As of extending the operator to be a generic one, we could argue the scope is still narrowed around providing the c.r.c-related services, not to be a single place for all outgoing traffic beyond the c.r.c. ingress.

As mentioned at the bottom of this doc (https://github.com/openshift/enhancements/pull/447/files#diff-74d208115f9ccd14e14cb6d90d466a9eR196) long term, we would like to see the generic solution that even insights-operator could use provided by OCP core.

For the API, would adding `v1alpha` in the endpoint help with the concerns anyhow?

Reply:
Taking a library approach solves a couple of the issues we're wanting to address:
1. abstract the mechanics of talking to cloudDot (http client configuration and uri stuff)
2. centralize the proxy configuration

It doesn't address gaining access to credentials necessary for authentication with cloudDot.  

Additionally, a library approach grants us the ability to just do it, which I _really_ like.

Reply:
I'm waiting for more feedback on this proposal from the OLM operators interested into this future.

Given we would go the library approach, where should the proxy configuration live?

Reply:
Can someone describe how the library would work with an Ansible based operator?
Would this be something we'd have to put into an Ansible collection? Also would like to understand the impact on EXD how the library would be built and made available for a certified operator release?  I think the documentation for the library would need some information on how an operator would gather the pull secret config plus any necessary RBAC role/cluster role needs.


But if all that is easy then I'm okay with a library approach as it might have a side effect that it doesn't need all the backporting into z streams.

I assume the library would also take into account the cluster global proxy configuration?

Reply:
For the ansible-operator, I can think of two ways (based more on how I think ansible-operator works, without much exprience):

1. explosing the library via executable that would be available in the operator image and used instead of the `curl` it uses today. You would need to figure out how to add the binary to your image. Disadvantage would be it would need to read configuration on every upload, or deal with caching.
2. running as a side-car container, exposing the local endpoint to provide the gathering functionality
Regarding the side-car: what do you think about this being the generic regardless it's ansible operator or not @smarterclayton, @jhjaggars? This way, we could also standardize on metrics and would actually feel like k8s way?

Reply:
What would be the advantages of using the library from the solution we have today - coded by ourselves? It seems that the long term solution of an upload service will better fix the air-gapped use cases, and would simplify credential management and RBAC configuration.

Reply:
At least two advantages that I can think of:
1. standardization of how to make the connection
2. clear instructions for users to follow

The sidecar container approach (basically a service wrapper around _the library_) should address the ansible operator case.   Would it be a big lift to include another image in the existing certification process for each operator?

Reply:
Another question: could the OLM operators create this bindings as part of the deployment process (or ask the administrator to do so?):

```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: some-olm-operator
  namespace: openshift-config
rules:
- apiGroups:
  - ""
  resources:
  - secrets
  resourceNames:
  - pull-secret
  - support
  verbs:
  - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: some-olm-operator
  namespace: openshift-config
roleRef:
  kind: Role
  name: some-olm-operator
subjects:
- kind: ServiceAccount
  name: operator
  namespace: some-olm-operator
```

If so, the sidecar could also pull the right token from the pull secret and watch for the changes.

The sidecar is probably my favorite one also as it could standardize on things like tracked metrics.